latch_config/config.py

```diff
@@ -51,17 +70,92 @@ class LoggingMode(str, Enum):
 T = TypeVar("T")
 
 
-def read_config(x: type[T], env_prefix: str = "") -> T:
+class _IsDataclass(Protocol):
+    __dataclass_fields__: ClassVar[dict[str, Field]]
+
+
+DC = TypeVar("DC", bound=_IsDataclass)
+
+
+class NoValue: ...
+
+
+class NestyError(Exception):
+    def __init__(self, msg: str, children: list[Self] | None = None):
+        self.msg = msg
+        self.children = children
+
+    def __str__(self):
+        if self.children is None:
+            return self.msg
+
+        res: list[str] = [self.msg]
+        for child in self.children:
+            res.append(indent(str(child), " -> "))
+
+        return "\n".join(res)
+
+
+def parse_config_val(t: type[T], v: Any) -> T:
+    if t is type(None):
+        if v is None:
+            return v
+
+        raise NestyError(f"value `{repr(v)}` is not assignable to `{t}`")
+
+    origin, args = get_origin(t), get_args(t)
+
+    if origin is None:
+        try:
+            if not isinstance(v, t):
+                raise TypeError(f"invalid literal `{v}` for primitive type `{t}`")
```

Comment:
```suggestion
                raise TypeError(f"expected a {t}: {v!r}")
```
for reference this is what data-validation does: https://github.com/latchbio/latch-data-validation/blob/80cd1c2b5742ad2a8688c0c7c19a3eaa5f217859/latch_data_validation/data_validation.py#L438-L461